sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala

```diff
@@ -2085,12 +2085,23 @@ class AstBuilder extends SqlBaseBaseVisitor[AnyRef] with SQLConfHelper with Logg
   private def canApplyRegex(ctx: ParserRuleContext): Boolean = withOrigin(ctx) {
     var parent = ctx.getParent
     while (parent != null) {
-      if (parent.isInstanceOf[NamedExpressionContext]) return true
+      if (parent.isInstanceOf[NamedExpressionContext] && isRegex(ctx.getText)) return true
       parent = parent.getParent
     }
     return false
   }
 
+  /**
+   * Returns whether the pattern is a regex expression (instead of a normal
+   * string). Normal string is a string with all alphabets/digits and "_".
+   */
+  private def isRegex(pattern: String): Boolean = {
+    for (p <- pattern) {
+      if (!Character.isLetterOrDigit(p) && p != '_') return true
+    }
```

Comment:
Maybe we can make it more like a scala code.
to use 
```
pattern.find()
or 
pattern.forall()
```

Reply:
done